=====================================
Part III: build your own workflow
=====================================

The task
----------

We want to perform read mapping and variant genotyping against a genome graph, a data structure which contains known genetic variation observed in a population or a species.

We will use `gramtools <https://github.com/iqbal-lab-org/gramtools>`_ to do this. It is not the fastest and it is not the most complete, but it is the one that Brice's PhD is all about.

Let's hope we don't find any bugs along the way :-)

The data
---------

We have 10 read sets from `Mycobacterium tuberculosis`, a bacterial pathogen. We also have its reference genome and a VCF containing known variants.


Desired workflow
-----------------

.. figure:: img/P3_dag.svg

    The steps of the workflow to implement.
    Two samples are shown here, but you have a total of 10 at your disposal.

.. tip::
    Start by using only two samples, as in the provided `config/samples.tsv`.

    Then when your workflow works with two samples, run on all 10 using:: 

        mv config/all_samples.tsv config/samples.tsv

Key information
----------------

Here is some key information to help you write the workflow.

First, check out the sketched Snakefile that will start you off:

.. literalinclude:: P3_dir/toComplete_Snakefile
        :language: python


