
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Build a full workflow &#8212; WMS_teaching 0.1 documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Going further" href="Further.html" />
    <link rel="prev" title="Build a full workflow" href="P3.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="build-a-full-workflow">
<h1>Build a full workflow<a class="headerlink" href="#build-a-full-workflow" title="Permalink to this headline">¶</a></h1>
<div class="section" id="the-task">
<h2>The task<a class="headerlink" href="#the-task" title="Permalink to this headline">¶</a></h2>
<p>We want to perform read mapping and variant genotyping against a genome graph, a data structure which contains known genetic variation observed in a population or a species.</p>
<p>We will use <a class="reference external" href="https://github.com/iqbal-lab-org/gramtools">gramtools</a> to do this, which Brice develops.</p>
</div>
<div class="section" id="the-data">
<h2>The data<a class="headerlink" href="#the-data" title="Permalink to this headline">¶</a></h2>
<p>We have 5 read sets from <cite>Mycobacterium tuberculosis</cite>, a bacterial pathogen. We also have its reference genome and a file describing known variants, in the <a class="reference external" href="https://github.com/samtools/hts-specs">VCF</a> format. The data files are in <cite>P3/data</cite>.</p>
<p>We will build a genome graph, map the reads to it, and find out which samples have which variants.</p>
</div>
<div class="section" id="desired-workflow">
<h2>Desired workflow<a class="headerlink" href="#desired-workflow" title="Permalink to this headline">¶</a></h2>
<div class="figure align-default" id="id4">
<img alt="../_images/P3_dag.svg" src="../_images/P3_dag.svg" /><p class="caption"><span class="caption-text">The steps of the workflow to implement. Only two samples are shown.</span><a class="headerlink" href="#id4" title="Permalink to this image">¶</a></p>
</div>
</div>
<div class="section" id="steps-explained">
<h2>Steps explained<a class="headerlink" href="#steps-explained" title="Permalink to this headline">¶</a></h2>
<table class="docutils align-center">
<colgroup>
<col style="width: 13%" />
<col style="width: 40%" />
<col style="width: 47%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Step</p></th>
<th class="head"><p>Purpose</p></th>
<th class="head"><p>Command</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference internal" href="#build"><span class="std std-ref">build</span></a></p></td>
<td><p>Build a genome graph from genome reference + variants</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">gramtools</span> <span class="pre">build</span> <span class="pre">{options}</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#quasimap"><span class="std std-ref">quasimap</span></a></p></td>
<td><p>Map reads from a sample to a genome graph</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">gramtools</span> <span class="pre">quasimap</span> <span class="pre">{options}</span></code></p></td>
</tr>
<tr class="row-even"><td><p>infer</p></td>
<td><p>Infer which variants the mapped sample has</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">gramtools</span> <span class="pre">infer</span> <span class="pre">{options}</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>compress</p></td>
<td><p>compress a VCF file and index it for fast access</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">bgzip</span> <span class="pre">{file}</span> <span class="pre">&amp;&amp;</span> <span class="pre">bcftools</span> <span class="pre">{file}.gz</span></code></p></td>
</tr>
<tr class="row-even"><td><p>mergeVcfs</p></td>
<td><p>combine compressed VCFs into one multi-sample VCF</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">bcftools</span> <span class="pre">merge</span> <span class="pre">-O</span> <span class="pre">z</span> <span class="pre">{inputs}</span> <span class="pre">&gt;</span> <span class="pre">{output}</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#plots"><span class="std std-ref">plots</span></a></p></td>
<td><p>make plots of the variation found (Brice’s script)</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">python3</span> <span class="pre">scripts/analyse_variants.py</span> <span class="pre">{in}.vcf.gz</span> <span class="pre">{outdir}</span></code></p></td>
</tr>
</tbody>
</table>
<p><strong>Get information the command options by running them on the command-line</strong>.</p>
<div class="section" id="build">
<span id="id1"></span><h3>build<a class="headerlink" href="#build" title="Permalink to this headline">¶</a></h3>
<p>This command takes a <cite>kmer-size</cite> option, reasonable values are &lt; 12.</p>
</div>
<div class="section" id="quasimap">
<span id="id2"></span><h3>quasimap<a class="headerlink" href="#quasimap" title="Permalink to this headline">¶</a></h3>
<p>This command takes a <cite>run-directory</cite> option, which stores all results for a given sample. It used again in the <code class="docutils literal notranslate"><span class="pre">infer</span></code> step.</p>
</div>
<div class="section" id="plots">
<span id="id3"></span><h3>plots<a class="headerlink" href="#plots" title="Permalink to this headline">¶</a></h3>
<p>The script plots, given a multi-sample VCF file, i) variant density and ii) the site frequency spectrum.</p>
<p>This script uses the excellent python package <a class="reference external" href="https://scikit-allel.readthedocs.io/en/stable/index.html">scikit-allel</a> for analysing genetic variation, if you’re interested!</p>
</div>
</div>
<div class="section" id="tips">
<h2>Tips<a class="headerlink" href="#tips" title="Permalink to this headline">¶</a></h2>
<div class="section" id="writing">
<h3>Writing<a class="headerlink" href="#writing" title="Permalink to this headline">¶</a></h3>
<p>Here is the Snakefile we start from:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">configfile</span><span class="p">:</span> <span class="s2">&quot;config/config.yaml&quot;</span>

<span class="n">samples</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;config/samples.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
	<span class="nb">input</span><span class="p">:</span>
		<span class="s2">&quot;</span><span class="si">{data}</span><span class="s2">/gramtools/build&quot;</span>

		<span class="c1">#expand(&quot;{results}/nuc_diversity.pdf&quot;,results=config[&quot;results&quot;]) # The final expected result</span>

<span class="n">rule</span> <span class="n">build</span><span class="p">:</span>
	<span class="nb">input</span><span class="p">:</span>
		<span class="n">ref</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">{data}</span><span class="s2">/TB_ref.fa&quot;</span><span class="p">,</span>
		<span class="n">vcf</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">{data}</span><span class="s2">/known_variants.vcf.gz&quot;</span>
	<span class="n">output</span><span class="p">:</span>
		<span class="n">directory</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{data}</span><span class="s2">/gramtools/build&quot;</span><span class="p">)</span>
	<span class="n">params</span><span class="p">:</span>
		<span class="c1">#k = ???</span>
	<span class="n">shell</span><span class="p">:</span>
		<span class="c1">#???</span>
</pre></div>
</div>
<p><strong>Keep updating the ``all`` rule with each step you add</strong>, so that you always have a functional workflow.</p>
</div>
<div class="section" id="data">
<h3>Data<a class="headerlink" href="#data" title="Permalink to this headline">¶</a></h3>
<p>Start by using only two samples, as in the provided <cite>config/samples.tsv</cite>.</p>
<p>Then when your workflow works with two samples, run on all of them using:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mv</span> <span class="n">config</span><span class="o">/</span><span class="n">all_samples</span><span class="o">.</span><span class="n">tsv</span> <span class="n">config</span><span class="o">/</span><span class="n">samples</span><span class="o">.</span><span class="n">tsv</span>
</pre></div>
</div>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">WMS_teaching</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../setup.html">Setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="../intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../P1.html">Part I: walkthrough</a></li>
<li class="toctree-l1"><a class="reference internal" href="../P2.html">Part II: fill in the gaps</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="P3.html">Part III: build a full workflow</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">Regular workflow</a></li>
<li class="toctree-l2"><a class="reference internal" href="Further.html">Going further</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../resources.html">Resources</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="P3.html">Build a full workflow</a><ul>
      <li>Previous: <a href="P3.html" title="previous chapter">Build a full workflow</a></li>
      <li>Next: <a href="Further.html" title="next chapter">Going further</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Brice Letcher & Paul Saary.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.3.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/P3/Workflow.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>