
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Build your own workflow &#8212; WMS_teaching 0.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Useful Resources" href="resources.html" />
    <link rel="prev" title="Fill in a workflow" href="P2.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="build-your-own-workflow">
<h1>Build your own workflow<a class="headerlink" href="#build-your-own-workflow" title="Permalink to this headline">¶</a></h1>
<div class="contents local topic" id="contents">
<ul class="simple">
<li><a class="reference internal" href="#the-task" id="id4">The task</a></li>
<li><a class="reference internal" href="#the-data" id="id5">The data</a></li>
<li><a class="reference internal" href="#desired-workflow" id="id6">Desired workflow</a></li>
<li><a class="reference internal" href="#key-information" id="id7">Key information</a><ul>
<li><a class="reference internal" href="#starting-point" id="id8">Starting point</a></li>
<li><a class="reference internal" href="#gramtools-commands" id="id9">Gramtools commands</a></li>
<li><a class="reference internal" href="#other-processes" id="id10">Other processes</a></li>
</ul>
</li>
<li><a class="reference internal" href="#bonus-points-improving-the-workflow" id="id11">Bonus Points: improving the workflow</a><ul>
<li><a class="reference internal" href="#logging" id="id12">Logging</a></li>
<li><a class="reference internal" href="#benchmarking" id="id13">Benchmarking</a></li>
<li><a class="reference internal" href="#restarts" id="id14">Restarts</a></li>
<li><a class="reference internal" href="#cluster" id="id15">Cluster</a></li>
</ul>
</li>
<li><a class="reference internal" href="#bonus-points-drug-resistance-prediction" id="id16">Bonus Points: drug resistance prediction</a></li>
</ul>
</div>
<div class="section" id="the-task">
<h2><a class="toc-backref" href="#id4">The task</a><a class="headerlink" href="#the-task" title="Permalink to this headline">¶</a></h2>
<p>We want to perform read mapping and variant genotyping against a genome graph, a data structure which contains known genetic variation observed in a population or a species.</p>
<p>We will use <a class="reference external" href="https://github.com/iqbal-lab-org/gramtools">gramtools</a> to do this. It is not the fastest and it is not the most complete, but it is the one that Brice’s PhD is all about.</p>
<p>Let’s hope we don’t find any bugs along the way :-)</p>
</div>
<div class="section" id="the-data">
<h2><a class="toc-backref" href="#id5">The data</a><a class="headerlink" href="#the-data" title="Permalink to this headline">¶</a></h2>
<p>We have 10 read sets from <cite>Mycobacterium tuberculosis</cite>, a bacterial pathogen. We also have its reference genome and a VCF containing known variants.
Have a look at the <cite>P3/data</cite> folder to see the data files.</p>
</div>
<div class="section" id="desired-workflow">
<h2><a class="toc-backref" href="#id6">Desired workflow</a><a class="headerlink" href="#desired-workflow" title="Permalink to this headline">¶</a></h2>
<div class="figure align-default" id="id3">
<img alt="_images/P3_dag.svg" src="_images/P3_dag.svg" /><p class="caption"><span class="caption-text">The steps of the workflow to implement.
Two samples are shown here, but you have a total of 10 at your disposal.</span><a class="headerlink" href="#id3" title="Permalink to this image">¶</a></p>
</div>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p>Start by using only two samples, as in the provided <cite>config/samples.tsv</cite>.</p>
<p>Then when your workflow works with two samples, run on all 10 using:</p>
<div class="last highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mv</span> <span class="n">config</span><span class="o">/</span><span class="n">all_samples</span><span class="o">.</span><span class="n">tsv</span> <span class="n">config</span><span class="o">/</span><span class="n">samples</span><span class="o">.</span><span class="n">tsv</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="key-information">
<h2><a class="toc-backref" href="#id7">Key information</a><a class="headerlink" href="#key-information" title="Permalink to this headline">¶</a></h2>
<p>Below is some key information to help you write the workflow.</p>
<div class="section" id="starting-point">
<h3><a class="toc-backref" href="#id8">Starting point</a><a class="headerlink" href="#starting-point" title="Permalink to this headline">¶</a></h3>
<p>Here is the Snakefile we start from:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>

<span class="n">configfile</span><span class="p">:</span> <span class="s2">&quot;config/config.yaml&quot;</span>

<span class="n">samples</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;config/samples.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="n">rule</span> <span class="nb">all</span><span class="p">:</span>
	<span class="nb">input</span><span class="p">:</span>
		<span class="n">expand</span><span class="p">(</span><span class="s2">&quot;{results}/nuc_diversity.pdf&quot;</span><span class="p">,</span><span class="n">results</span><span class="o">=</span><span class="n">config</span><span class="p">[</span><span class="s2">&quot;results&quot;</span><span class="p">])</span>

<span class="n">rule</span> <span class="n">build</span><span class="p">:</span>
	<span class="nb">input</span><span class="p">:</span>
		<span class="n">ref</span><span class="o">=</span><span class="s2">&quot;{data}/TB_ref.fa&quot;</span><span class="p">,</span>
		<span class="n">vcf</span><span class="o">=</span><span class="s2">&quot;{data}/known_variants.vcf.gz&quot;</span>
	<span class="n">output</span><span class="p">:</span>
		<span class="n">directory</span><span class="p">(</span><span class="s2">&quot;{data}/gramtools/build&quot;</span><span class="p">)</span>
	<span class="n">params</span><span class="p">:</span>
		<span class="c1">#?</span>
	<span class="n">singularity</span><span class="p">:</span>
		<span class="n">config</span><span class="p">[</span><span class="s2">&quot;?&quot;</span><span class="p">]</span>
	<span class="n">shell</span><span class="p">:</span>
		<span class="c1">#?</span>
</pre></div>
</div>
</div>
<div class="section" id="gramtools-commands">
<h3><a class="toc-backref" href="#id9">Gramtools commands</a><a class="headerlink" href="#gramtools-commands" title="Permalink to this headline">¶</a></h3>
<p>Get information on available gramtools commands by running <code class="docutils literal notranslate"><span class="pre">singularity</span> <span class="pre">exec</span> <span class="pre">P3/config/WMS.img</span> <span class="pre">gramtools</span></code></p>
<p>Tips:</p>
<ul class="simple">
<li><code class="docutils literal notranslate"><span class="pre">build</span></code> takes a <cite>kmer-size</cite> option, reasonable values are &lt; 12.</li>
<li>gramtools commands after <code class="docutils literal notranslate"><span class="pre">build</span></code> use a <cite>run directory</cite>, which you create in the <code class="docutils literal notranslate"><span class="pre">quasimap</span></code> phase. This
directory should contain all results for a given sample.</li>
</ul>
</div>
<div class="section" id="other-processes">
<h3><a class="toc-backref" href="#id10">Other processes</a><a class="headerlink" href="#other-processes" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>In process <strong>mergeVcfs</strong>, we need to combine the vcfs from <code class="docutils literal notranslate"><span class="pre">infer</span></code> into one vcf using the program <code class="docutils literal notranslate"><span class="pre">bcftools</span></code>.
<code class="docutils literal notranslate"><span class="pre">bcftools</span></code> requires compressed and indexed vcfs as input. You can compress using <code class="docutils literal notranslate"><span class="pre">bgzip</span></code>.</li>
<li>The process <strong>plots</strong> uses the python3 script in P3/scripts/analyse_variant.py. Try running it on its own to see options.
The script uses the excellent python package
<a class="reference external" href="https://scikit-allel.readthedocs.io/en/stable/index.html">scikit-allel</a> for analysing genetic variation, if you’re interested!</li>
</ul>
</div>
</div>
<div class="section" id="bonus-points-improving-the-workflow">
<h2><a class="toc-backref" href="#id11">Bonus Points: improving the workflow</a><a class="headerlink" href="#bonus-points-improving-the-workflow" title="Permalink to this headline">¶</a></h2>
<p>Use the Snakemake readthedocs webpage for help!</p>
<div class="section" id="logging">
<h3><a class="toc-backref" href="#id12">Logging</a><a class="headerlink" href="#logging" title="Permalink to this headline">¶</a></h3>
<p>Add logging to one or more rules, to capture stdout and stderr in files.
<a class="reference external" href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#">relevant doc</a></p>
</div>
<div class="section" id="benchmarking">
<h3><a class="toc-backref" href="#id13">Benchmarking</a><a class="headerlink" href="#benchmarking" title="Permalink to this headline">¶</a></h3>
<p>Snakemake can measure CPU/wall clock time and RAM use of each rule.
Find out how, and try it out on a rule.</p>
</div>
<div class="section" id="restarts">
<h3><a class="toc-backref" href="#id14">Restarts</a><a class="headerlink" href="#restarts" title="Permalink to this headline">¶</a></h3>
<p>What does Snakemake rely on to know where in the DAG to restart after a failed run?</p>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">Search the doc for <code class="docutils literal notranslate"><span class="pre">timestamps</span></code>.</p>
</div>
<ul class="simple">
<li>Try modifying a file yourself such that Snakemake will re-run the workflow from rule <cite>mergeVcfs</cite>.</li>
<li>Find out the command-line option to re-run the workflow from any user-specified rule. <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/project_info/faq.html">relevant doc</a></li>
</ul>
</div>
<div class="section" id="cluster">
<h3><a class="toc-backref" href="#id15">Cluster</a><a class="headerlink" href="#cluster" title="Permalink to this headline">¶</a></h3>
<p>Figure out how to submit the workflow to the cluster. Note that cluster parameters should not go in the workflow itself, otherwise it is no longer independent of where it is run.
<a class="reference external" href="https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html#">relevant doc</a></p>
<p>For testing, ask us if there is cluster access, or if we can run it for you on EBI cluster (which uses LSF).</p>
</div>
</div>
<div class="section" id="bonus-points-drug-resistance-prediction">
<h2><a class="toc-backref" href="#id16">Bonus Points: drug resistance prediction</a><a class="headerlink" href="#bonus-points-drug-resistance-prediction" title="Permalink to this headline">¶</a></h2>
<p>The dataset contains at least one sample which is resistant to a drug against tuberculosis (TB).
Can you find which samples are resistant to which known TB drugs?</p>
<p>You can use the <a class="reference external" href="https://github.com/mykrobe-tools/mykrobe">mykrobe</a> program to do this. It is available in
the singularity image:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">singularity</span> <span class="n">exec</span> <span class="n">P3</span><span class="o">/</span><span class="n">WMS</span><span class="o">.</span><span class="n">img</span> <span class="n">mykrobe</span>
</pre></div>
</div>
<p>Check the drug resistance predictions by <cite>mykrobe</cite> are present in the VCFs you produced using the workflow.</p>
<p>Running mykrobe and making a report could be added to the workflow.</p>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">WMS_teaching</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="setup.html">Setup</a></li>
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="P1.html">Part I: walkthrough</a></li>
<li class="toctree-l1"><a class="reference internal" href="P2.html">Part II: fill in the gaps</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Part III: build a full workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="resources.html">Resources</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="P2.html" title="previous chapter">Fill in  a workflow</a></li>
      <li>Next: <a href="resources.html" title="next chapter">Useful Resources</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Brice Letcher & Paul Saary.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/P3.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>